[SPARK-43313][SQL] Adding missing column DEFAULT values for MERGE INSERT actions

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR updates column DEFAULT assignment to add missing values for MERGE INSERT actions. This brings the behavior to parity with non-MERGE INSERT commands.

* It also adds a small convenience feature where if the provided default value is a literal of a wider type than the target column, but the literal value fits within the narrower type, just coerce it for convenience. For example, `CREATE TABLE t (col INT DEFAULT 42L)` returns an error before this PR because `42L` has a long integer type which is wider than `col`, but after this PR we just coerce it to `42` since the value fits within the short integer range.
* We also add the `SupportsCustomSchemaWrite` interface which tables may implement to exclude certain pseudocolumns from consideration when resolving column DEFAULT values.

### Why are the changes needed?

These changes make column DEFAULT values more usable in more types of situations.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds new unit test coverage.

dtenedor committed:
commit

dtenedor committed:
fix tests

dtenedor committed:
fix tests

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:
```diff
@@ -344,3 +358,13 @@ object ResolveDefaultColumns {
+ * The 'ignoreColumnsForInserts' field represents a set of column names to exclude from
+ * consideration when performing INSERT INTO commands, if any.
+ */
+trait TableWithRawSchemaForColumnDefaults {
```
Should this be a mixin trait in DSV2?

dtenedor replied:
We discussed offline, it's probably more straightforward to have the separate trait that we only have to think about when it's relevant for specific sources than in the core DSV2 API.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:
```diff
@@ -344,3 +358,13 @@ object ResolveDefaultColumns {
+
+/**
+ * Represents a table with an underlying raw schema to use for resolving DEFAULT column references.
+ * The 'ignoreColumnsForInserts' field represents a set of column names to exclude from
```
Let's move(or duplicate) the comment to the method

dtenedor replied:
Good idea, done.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:
```diff
@@ -344,3 +358,13 @@ object ResolveDefaultColumns {
+ * consideration when performing INSERT INTO commands, if any.
+ */
+trait TableWithRawSchemaForColumnDefaults {
+  def rawSchema: Option[StructType]
```
Why the result is optional? Could you write an example about the difference between `rawSchema` and `schema`?

dtenedor replied:
Good point, I changed this to become non-optional an added a comment with an example of how this might differ from the regular schema.

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor committed:
fix casting bug

gengliangwang commented on sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:
```diff
@@ -1044,7 +1044,7 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
     object Errors {
       val COMMON_SUBSTRING = " has a DEFAULT value"
-      val COLUMN_DEFAULT_NOT_FOUND = "`default` cannot be resolved."
+      val COLUMN_DEFAULT_NOT_FOUND = "The table or view `t` cannot be found"
```
The error message doesn't seems related to COLUMN_DEFAULT_NOT_FOUND

dtenedor replied:
Fixed.

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:
```diff
@@ -343,4 +363,11 @@ object ResolveDefaultColumns {
     }
   }
+
+  trait CustomInsertSchema {
```
Move this under package `org.apache.spark.sql.connector.write` as `SupportsCustomSchemaWrite`? 
We will need the implement a new class for testing the new trait(interface) as well

dtenedor replied:
@gengliangwang it needs to be visible from the `ResolveDefaultColumns` rule which is under `package org.apache.spark.sql.catalyst.analysis`. That `package org.apache.spark.sql.connector.write` is not visible in this way. Is there a different package we should use instead?

I will add a test table that inherits this trait, and exercise it in the unit tests and comment here again when ready.

gengliangwang replied:
It should be accessible. 
https://github.com/apache/spark/tree/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/write

dtenedor replied:
Thanks for the pointer, I was able to implement this suggestion.

gengliangwang commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/SupportsCustomSchemaWrite.java:
```diff
@@ -0,0 +1,37 @@
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * Trait for tables that support custom schemas for certain write operations.
```
Let's make it more specified about the "certain write operations". For example, insertion with default values.

dtenedor replied:
Thanks, that's a good suggestion. I updated the comment accordingly to make it better.

gengliangwang commented on sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:
```diff
@@ -1096,7 +1096,7 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
       assert(intercept[AnalysisException] {
         sql("insert into t values(false, default)")
-      }.getMessage.contains(Errors.COLUMN_DEFAULT_NOT_FOUND))
+      }.getMessage.contains(Errors.TARGET_TABLE_NOT_FOUND))
```
Why the error becomes table not found from column default not found?

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:
```diff
@@ -218,10 +219,29 @@ object ResolveDefaultColumns {
+      // If the provided default value is a literal of a wider type than the target column, but the
+      // literal value fits within the narrower type, just coerce it for convenience.
+      val result = if (analyzed.isInstanceOf[Literal] &&
+        !Seq(dataType, analyzed.dataType).exists(_ == BooleanType)) {
```
Add a comment about why can't any of the data types being Boolean?

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:
```diff
@@ -218,10 +219,29 @@ object ResolveDefaultColumns {
+            None
+          }
+        } catch {
+          case _: SparkThrowable | _: RuntimeException => None
```
If it is TryCast, why would we need to use `try...cast` here? Shall we simply check if the cast is `resolved` before eval?

gengliangwang commented:
@dtenedor BTW could you mention the new trait in the PR description?

dtenedor reviewed (commented):
Thanks for your review!

dtenedor committed:
respond to code review comments

gengliangwang commented:
@dtenedor FYI there are test failures in the latest code.

gengliangwang commented:
cc @cloud-fan since this is DSV2 related.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:
```diff
@@ -589,8 +631,10 @@ case class ResolveDefaultColumns(
         r.tableMeta.schema
-      case d: DataSourceV2Relation if !d.skipSchemaResolution && !d.isStreaming =>
-        CatalogV2Util.v2ColumnsToStructType(d.table.columns())
+      case DataSourceV2Relation(table: SupportsCustomSchemaWrite, _, _, _, _) =>
```
We can create follow-up jira to cover the resolution of insertion with SupportsCustomSchemaWrite.

dtenedor replied:
👍  I created it here: https://issues.apache.org/jira/browse/SPARK-43358

dtenedor committed:
fix test

dtenedor commented:
> @dtenedor FYI there are test failures in the latest code.

@gengliangwang thanks, fixed

gengliangwang commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/SupportsCustomSchemaWrite.java:
```diff
@@ -0,0 +1,38 @@
+ * Trait for tables that support custom schemas for write operations including INSERT INTO commands
+ * whose target table columns have explicit or implicit default values.
+ *
+ * @since 3.4.0
```
3.4.1?

dtenedor replied:
Done.

dtenedor committed:
Update SupportsCustomSchemaWrite.java

gengliangwang commented:
Thanks, merging to master/branch-3.4

pan3793 commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/SupportsCustomSchemaWrite.java:
```diff
@@ -0,0 +1,38 @@
+     *
+     * @return the new schema to use for this process.
+     */
+    StructType customSchemaForInserts();
```
Looks schema is deprecated in SPARK-42398, and `Column[]` should be used instead

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/SupportsCustomSchemaWrite.java:
```diff
@@ -0,0 +1,38 @@
+ * Trait for tables that support custom schemas for write operations including INSERT INTO commands
+ * whose target table columns have explicit or implicit default values.
+ *
+ * @since 3.4.1
```
I'm reverting this from branch 3.4 as we shouldn't add new API during backporting.

dongjoon-hyun commented:
Thank you for reverting.
